### PR TITLE
PWN-7169 - same token error, price impact, amount error

### PR DESCRIPTION
**Pull request by NikolaiEmelianov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7169

## Description of Work

same token error, price impact, amount error

### CODE REVIEW

**Comment by 1sakov** [comment]
@NikolaiEmelianov why do you have so much commits?

**Comment by NikolaiEmelianov** [comment]
> 

https://github.com/p2p-org/key-app-android/pull/1309 based on this local branch, "Squash and merge" delete history

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/price_impact/SwapPriceImpact.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package org.p2p.wallet.swap.jupiter.statemanager.price_impact
+
+import java.math.BigDecimal
+import org.p2p.core.utils.isLessThan
+
+enum class SwapPriceImpact {
+
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/price_impact/SwapPriceImpact.kt:7]: remove space

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/price_impact/SwapPriceImpact.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    NORMAL, YELLOW, RED
+}
+
+private val threePercent
+    get() = BigDecimal.valueOf(0.3)
+
+private val onePercent
+    get() = BigDecimal.valueOf(0.1)
+
+fun BigDecimal.toPriceImpactType(): SwapPriceImpact {
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/price_impact/SwapPriceImpact.kt:17]: not obvious, I suggest to move this method to SwapPriceImpact

Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/price_impact/SwapPriceImpact.kt:17]: SwapPriceImpact is enum, not support companion object. moved to presenter

Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/price_impact/SwapPriceImpact.kt]: `SwapPriceImpact.parseValue(impact: BigDecimal)`

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt [hunk 3]
```diff
@@ -160,133 +165,223 @@ class JupiterSwapPresenter(
 
     private fun handleFeatureState(state: SwapState) {
         cancelRateJobs()
+        featureState = state
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:168]: featureState -> currentFeatureState

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt [hunk 4]
```diff
@@ -160,133 +165,223 @@ class JupiterSwapPresenter(
 
     private fun handleFeatureState(state: SwapState) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:166]: handleNewFeatureState / handleFeatureStateUpdate

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt [hunk 5]
```diff
@@ -160,133 +165,223 @@ class JupiterSwapPresenter(
...
+            is SwapFeatureException.RoutesNotFound ->
+                view?.setButtonState(buttonState = buttonMapper.mapRouteNotFound())
+            is SwapFeatureException.NotValidTokenA -> {
+                val tokenA = state.previousFeatureState.getTokensPair().first
+                this.widgetAState = widgetMapper.mapErrorTokenAAmount(
+                    tokenA = tokenA,
+                    oldWidgetAState = widgetAState,
+                    notValidAmount = featureException.notValidAmount
+                )
+                if (tokenA != null) getRateTokenA(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:196]: {}

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt [hunk 6]
```diff
@@ -160,133 +165,223 @@ class JupiterSwapPresenter(
...
-        widgetBState = widgetMapper.mapWidgetLoading(tokenType = SwapTokenType.TOKEN_B)
+    private fun handleInitialLoading(state: SwapState.InitialLoading) {
+        mapWidgetStates(state)
         updateWidgets()
         view?.setButtonState(buttonState = SwapButtonState.Hide)
     }
 
-    private fun getRateTokenA(tokenA: SwapTokenModel, tokenAmount: BigDecimal) {
+    private fun mapWidgetStates(state: SwapState): Pair<SwapWidgetModel, SwapWidgetModel> {
+        val result = when (state) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:245]: explicit type please

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt [hunk 7]
```diff
@@ -160,133 +165,223 @@ class JupiterSwapPresenter(
...
+        return result
+    }
+
+    private fun getRateTokenA(widgetAModel: SwapWidgetModel, tokenA: SwapTokenModel, tokenAmount: BigDecimal) {
         rateTokenAJob?.cancel()
         rateTokenAJob = rateLoaderTokenA.getRate(tokenA)
             .flowOn(dispatchers.io)
-            .onEach { handleRateLoader(state = it, tokenType = SwapTokenType.TOKEN_A, tokenAmount = tokenAmount) }
+            .onEach {
+                if (isActive) handleRateTokenALoader(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:291]: {}

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt [hunk 8]
```diff
@@ -160,133 +165,223 @@ class JupiterSwapPresenter(
...
+        widgetBState = result.second
+        return result
+    }
+
+    private fun getRateTokenA(widgetAModel: SwapWidgetModel, tokenA: SwapTokenModel, tokenAmount: BigDecimal) {
         rateTokenAJob?.cancel()
         rateTokenAJob = rateLoaderTokenA.getRate(tokenA)
             .flowOn(dispatchers.io)
-            .onEach { handleRateLoader(state = it, tokenType = SwapTokenType.TOKEN_A, tokenAmount = tokenAmount) }
+            .onEach {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:290]: maybe add filter?

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt [hunk 9]
```diff
@@ -160,133 +165,223 @@ class JupiterSwapPresenter(
...
+        widgetAState = newWidgetModel
+        view?.setFirstTokenWidgetState(state = widgetAState)
+    }
+
+    private fun handleRateTokenBLoader(
+        widgetBModel: SwapWidgetModel,
+        state: SwapRateLoaderState,
+        tokenAmount: BigDecimal,
+    ) {
+
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:333]:
```suggestion
```

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt [hunk 10]
```diff
@@ -160,133 +165,223 @@ class JupiterSwapPresenter(
...
 
     private fun updateWidgets() {
         view?.setFirstTokenWidgetState(state = widgetAState)
         view?.setSecondTokenWidgetState(state = widgetBState)
     }
+
+    private val threePercent
+        get() = BigDecimal.valueOf(0.3)
+
+    private val onePercent
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:377]: fields should be placed in top

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapButtonMapper.kt [hunk 11]
```diff
@@ -13,6 +13,17 @@ class SwapButtonMapper {
...
+    fun mapRouteNotFound(): SwapButtonState.Disabled =
+        SwapButtonState.Disabled(TextContainer(R.string.swap_main_button_route_not_found))
+
+    fun mapSameToken(): SwapButtonState.Disabled =
+        SwapButtonState.Disabled(TextContainer(R.string.swap_main_button_route_not_found))
+
+    fun mapTokenAmountNotEnough(tokenA: SwapTokenModel?): SwapButtonState.Disabled =
+        tokenA?.tokenSymbol?.let { symbol ->
+            SwapButtonState.Disabled(TextContainer(R.string.swap_main_button_not_enough_amount, symbol))
+        } ?: mapEnterAmount()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapButtonMapper.kt:25]:
```suggestion
        tokenA?.tokenSymbol
            ?.let { SwapButtonState.Disabled(TextContainer(R.string.swap_main_button_not_enough_amount, it )) } 
            ?: mapEnterAmount()
```

**Review by gslevinkov — APPROVED** [review]

**Review by 1sakov — APPROVED** [review]

**Pull request merged**